Add multistep DPM-Solver discrete scheduler

LuChengTHU opened the pull request:
Add the multistep version of DPM-Solver, accepting discrete time steps as DDPM, DDIM, and PNDM.

The solver can successfully run with the stable-diffusion pipeline for both pytorch and jax versions.

Currently, it supports the following algorithms:

- multistep DPM-Solver (for discretizing the integral w.r.t. noise prediction model) with order = 1, 2, 3
- multistep DPM-Solver++ (for discretizing the integral w.r.t. data prediction model) with order = 1, 2, 3

For more details of the algorithms, please refer to https://github.com/LuChengTHU/dpm-solver

#1101

patil-suraj commented on src/diffusers/__init__.py:
```diff
@@ -42,6 +42,7 @@
     from .schedulers import (
         DDIMScheduler,
         DDPMScheduler,
+        DPMSolverDiscreteScheduler,
```
Very cool! we should also add this in the docs in 
`docs/source/api/schedulers.mdx`

LuChengTHU replied:
I've added the auto docs in commit f65012c

patil-suraj commented on src/diffusers/schedulers/scheduling_dpmsolver_discrete.py:
```diff
@@ -0,0 +1,471 @@
+
+
+class DPMSolverDiscreteScheduler(SchedulerMixin, ConfigMixin):
+    """
+    DPM-Solver (and the improved version DPM-Solver++) is a fast dedicated high-order solver for diffusion ODEs with
+    the convergence order guarantee. Empirically, sampling by DPM-Solver with only 20 steps can generate high-quality
+    samples, and it can generate quite good samples even in only 10 steps.
+
+    For more details, see the original paper: https://arxiv.org/abs/2206.00927 and https://arxiv.org/abs/2211.01095
+
+    Currently, we support the multistep DPM-Solver for both noise prediction models and data prediction models. We
+    recommend to use `solver_order=2` for guided sampling, and `solver_order=3` for unconditional sampling.
+
+    We also support the "dynamic thresholding" method in Imagen (https://arxiv.org/abs/2205.11487). For pixel-space
+    diffusion models, you can set both `predict_x0=True` and `thresholding=True` to use the dynamic thresholding. Note
+    that the thresholding method is unsuitable for latent-space diffusion models (such as stable-diffusion).
+
+    [`~ConfigMixin`] takes care of storing all config attributes that are passed in the scheduler's `__init__`
+    function, such as `num_train_timesteps`. They can be accessed via `scheduler.config.num_train_timesteps`.
+    [`~ConfigMixin`] also provides general loading and saving functionality via the [`~ConfigMixin.save_config`] and
+    [`~ConfigMixin.from_config`] functions.
```
Really nice doc!

patil-suraj commented on src/diffusers/schedulers/scheduling_dpmsolver_discrete.py:
```diff
@@ -0,0 +1,471 @@
+        self.init_noise_sigma = 1.0
+
+        # settings for DPM-Solver
+        self.solver_order = solver_order
+        self.predict_x0 = predict_x0
+        self.thresholding = thresholding
+        self.sample_max_value = sample_max_value
+        self.denoise_final = denoise_final
```
It's not necessary to save these values here , all the args in `__init__` can be accessed using `self.config`

LuChengTHU replied:
resolved in commit 588ce52

patil-suraj commented on src/diffusers/schedulers/scheduling_dpmsolver_discrete.py:
```diff
@@ -0,0 +1,471 @@
+            .astype(np.int64)
+        )
+        self.timesteps = torch.from_numpy(timesteps).to(device)
+        self.model_outputs = [
+            None,
+        ] * self.solver_order
```
```suggestion
        self.model_outputs = [None] * self.solver_order
```

LuChengTHU replied:
the code style checker suggests the original code style, so I use the original style :)

patil-suraj commented on src/diffusers/schedulers/scheduling_dpmsolver_discrete.py:
```diff
@@ -0,0 +1,471 @@
+            x0_pred = (sample - sigma_t * model_output) / alpha_t
+            if self.thresholding:
+                # Dynamic thresholding in https://arxiv.org/abs/2205.11487
+                p = 0.995  # A hyperparameter in the paper of "Imagen" (https://arxiv.org/abs/2205.11487).
```
Is this a fixed value or can be changed ? If the latter, then maybe let's make it `init` arg so it can be configured by the user.

LuChengTHU replied:
resolved in commit 588ce52, add in the init

patil-suraj commented on src/diffusers/schedulers/scheduling_dpmsolver_discrete.py:
```diff
@@ -0,0 +1,471 @@
+            if self.thresholding:
+                # Dynamic thresholding in https://arxiv.org/abs/2205.11487
+                p = 0.995  # A hyperparameter in the paper of "Imagen" (https://arxiv.org/abs/2205.11487).
+                s = torch.quantile(torch.abs(x0_pred).reshape((x0_pred.shape[0], -1)), p, dim=1)
+                s = torch.maximum(s, self.sample_max_value * torch.ones_like(s).to(s.device))[
+                    (...,) + (None,) * (x0_pred.ndim - 1)
+                ]
+                x0_pred = torch.clamp(x0_pred, -s, s) / s
```
(nit)
For readability let's not use single letter variable

LuChengTHU replied:
resolved in commit 588ce52, change the name

patil-suraj commented on src/diffusers/schedulers/scheduling_dpmsolver_discrete.py:
```diff
@@ -0,0 +1,471 @@
+        Returns:
+            `torch.FloatTensor`: the sample tensor at the previous timestep.
+        """
+        t, s0, s1 = prev_timestep, timestep_list[-1], timestep_list[-2]
+        m0, m1 = model_output_list[-1], model_output_list[-2]
+        lambda_t, lambda_s0, lambda_s1 = self.lambda_t[t], self.lambda_t[s0], self.lambda_t[s1]
+        alpha_t, alpha_s0 = self.alpha_t[t], self.alpha_t[s0]
+        sigma_t, sigma_s0 = self.sigma_t[t], self.sigma_t[s0]
+        h, h_0 = lambda_t - lambda_s0, lambda_s0 - lambda_s1
+        r0 = h_0 / h
+        D0, D1 = m0, (1.0 / r0) * (m0 - m1)
+        if self.predict_x0:
+            if self.solver_type == "dpm_solver":
+                x_t = (
+                    (sigma_t / sigma_s0) * sample
+                    - (alpha_t * (torch.exp(-h) - 1.0)) * D0
+                    - 0.5 * (alpha_t * (torch.exp(-h) - 1.0)) * D1
+                )
+            elif self.solver_type == "taylor":
+                x_t = (
+                    (sigma_t / sigma_s0) * sample
+                    - (alpha_t * (torch.exp(-h) - 1.0)) * D0
+                    + (alpha_t * ((torch.exp(-h) - 1.0) / h + 1.0)) * D1
+                )
+        else:
+            if self.solver_type == "dpm_solver":
+                x_t = (
+                    (alpha_t / alpha_s0) * sample
+                    - (sigma_t * (torch.exp(h) - 1.0)) * D0
+                    - 0.5 * (sigma_t * (torch.exp(h) - 1.0)) * D1
+                )
+            elif self.solver_type == "taylor":
+                x_t = (
+                    (alpha_t / alpha_s0) * sample
+                    - (sigma_t * (torch.exp(h) - 1.0)) * D0
+                    - (sigma_t * ((torch.exp(h) - 1.0) / h - 1.0)) * D1
+                )
+        return x_t
```
Very cool! 

(nit)
Would be nice if we point to the equations in the paper corresponding to this code block. This will make it easy for the users who are wants to go through the paper and implementation.

LuChengTHU replied:
resolved in commit 588ce52, add comments

patil-suraj commented on src/diffusers/schedulers/scheduling_dpmsolver_discrete.py:
```diff
@@ -0,0 +1,471 @@
+        denoise_second = (step_index == len(self.timesteps) - 2) and self.denoise_final
+
+        model_output = self.convert_model_output(model_output, timestep, sample)
+        for i in range(self.solver_order - 1):
+            self.model_outputs[i] = self.model_outputs[i + 1]
+        self.model_outputs[-1] = model_output
+
+        if self.solver_order == 1 or self.lower_order_nums < 1 or denoise_final:
+            prev_sample = self.dpm_solver_first_order_update(model_output, timestep, prev_timestep, sample)
+        elif self.solver_order == 2 or self.lower_order_nums < 2 or denoise_second:
+            timestep_list = [self.timesteps[step_index - 1], timestep]
+            prev_sample = self.multistep_dpm_solver_second_order_update(
+                self.model_outputs, timestep_list, prev_timestep, sample
+            )
+        else:
+            timestep_list = [self.timesteps[step_index - 2], self.timesteps[step_index - 1], timestep]
+            prev_sample = self.multistep_dpm_solver_third_order_update(
+                self.model_outputs, timestep_list, prev_timestep, sample
+            )
+
+        if self.lower_order_nums < self.solver_order:
+            self.lower_order_nums += 1
```
Love this API ❤️

patil-suraj commented on src/diffusers/schedulers/scheduling_dpmsolver_discrete_flax.py:
```diff
@@ -0,0 +1,554 @@
+    state: DPMSolverDiscreteSchedulerState
+
+
+class FlaxDPMSolverDiscreteScheduler(FlaxSchedulerMixin, ConfigMixin):
```
Really cool that you added the flax version! 

Pretty much the same comments here as the torch scheduler

LuChengTHU replied:
resolved all the modifications as in pytorch version

patil-suraj commented on tests/test_config.py:
```diff
@@ -283,3 +290,15 @@ def test_load_pndm(self):
+        logger = logging.get_logger("diffusers.configuration_utils")
+
+        with CaptureLogger(logger) as cap_logger:
+            pndm = DPMSolverDiscreteScheduler.from_config(
```
```suggestion
            dpm = DPMSolverDiscreteScheduler.from_config(
```

patil-suraj commented on tests/test_config.py:
```diff
@@ -283,3 +290,15 @@ def test_load_pndm(self):
+                "hf-internal-testing/tiny-stable-diffusion-torch", subfolder="scheduler"
+            )
+
+        assert pndm.__class__ == DPMSolverDiscreteScheduler
```
```suggestion
        assert dpm.__class__ == DPMSolverDiscreteScheduler
```

patil-suraj commented on tests/test_scheduler.py:
```diff
@@ -549,6 +550,173 @@ def test_full_loop_with_no_set_alpha_to_one(self):
         assert abs(result_mean.item() - 0.1941) < 1e-3
 
 
+class DPMSolverDiscreteSchedulerTest(SchedulerCommonTest):
+    scheduler_classes = (DPMSolverDiscreteScheduler,)
+    forward_default_kwargs = (("num_inference_steps", 25),)
+
+    def get_scheduler_config(self, **kwargs):
+        config = {
+            "num_train_timesteps": 1000,
+            "beta_start": 0.0001,
+            "beta_end": 0.02,
+            "beta_schedule": "linear",
+            "solver_order": 2,
+            "predict_x0": True,
+            "thresholding": False,
+            "sample_max_value": 1.0,
+            "solver_type": "dpm_solver",
+            "denoise_final": False,
+        }
+
+        config.update(**kwargs)
+        return config
+
+    def check_over_configs(self, time_step=0, **config):
+        kwargs = dict(self.forward_default_kwargs)
+        num_inference_steps = kwargs.pop("num_inference_steps", None)
+        sample = self.dummy_sample
+        residual = 0.1 * sample
+        dummy_past_residuals = [residual + 0.2, residual + 0.15, residual + 0.10]
+
+        for scheduler_class in self.scheduler_classes:
+            scheduler_config = self.get_scheduler_config(**config)
+            scheduler = scheduler_class(**scheduler_config)
+            scheduler.set_timesteps(num_inference_steps)
+            # copy over dummy past residuals
```
Thank you for adding the tests !

patil-suraj reviewed (approved):
Great work @LuChengTHU, super cool PR!

The PR already looks super good, I just left some nits. 

Also tried it out with stable diffusion and it works really well, and generates really good output with 20-25 steps, for some examples even 10-15 🤯 !

QQ: What are the recommended setting for stable diffusion, I used the following and it worked super well. We will need to update these in the config of sd checkpoints.

```python
dpm = DPMSolverDiscreteScheduler.from_config("CompVis/stable-diffusion-v1-4", subfolder="scheduler", solver_order=2, predict_x0=False, denoise_final=True)
```

And is this scheduler different from the DPM version which requires two model evaluation per step ?  

We are still thinking about an API for schedulers that require two model evaluations. Will have a draft PR next week. Then we can add the other version of DPM once the API is finalised. 

Also, to differentiate it from the other scheduler could we call this `DPMMultiStepScheduler` ?

LuChengTHU commented:
> Great work @LuChengTHU, super cool PR!
> 
> The PR already looks super good, I just left some nits.
> 
> Also tried it out with stable diffusion and it works really well, and generates really good output with 20-25 steps, for some examples even 10-15 🤯 !
> 
> QQ: What are the recommended setting for stable diffusion, I used the following and it worked super well. We will need to update these in the config of sd checkpoints.
> 
> ```python
> dpm = DPMSolverDiscreteScheduler.from_config("CompVis/stable-diffusion-v1-4", subfolder="scheduler", solver_order=2, predict_x0=False, denoise_final=True)
> ```
> 
> And is this scheduler different from the DPM version which requires two model evaluation per step ?
> 
> We are still thinking about an API for schedulers that require two model evaluations. Will have a draft PR next week. Then we can add the other version of DPM once the API is finalised.
> 
> Also, to differentiate it from the other scheduler could we call this `DPMMultiStepScheduler` ?

Thank you very much for the quick and high-quality review!

Q1: What are the recommended settings for stable diffusion?

In my DPM-Solver++ paper, I use the following setting which works similar to your settings:
```
dpm = DPMSolverDiscreteScheduler.from_config(
    "CompVis/stable-diffusion-v1-4",  # or use the v1-5 version
    subfolder="scheduler",
    solver_order=2,
    predict_x0=True,
    thresholding=False,
    solver_type="dpm_solver",
    denoise_final=False,  # can also be True, but the influence is small
)
```

Q2: is this scheduler different from the DPM version, which requires two model evaluations per step?

Yes, they are different. The DPM version you mentioned is actually the singlestep DPM-Solver in my repo, which is proposed by my paper "DPM-Solver". The currently committed PR version is the multistep DPM-Solver in my repo, which is proposed by my other paper "DPM-Solver++".

In the paper "DPM-Solver++", I carefully compare the different settings of the solvers for **guided sampling** (i.e. conditional sampling) by diffusion models, and I find that:

0. All previous high-order solvers are unstable and fail to generate samples for large guidance scales and small steps (<20).

1. We further proposed algorithms that use `predict_x0=True`, i.e., discretizing the integral w.r.t. the data prediction model, and can greatly stabilize the sample quality for small steps (<20).

2. We find that for large guidance scales, higher-order solvers are extremely unstable. So we only use the 2nd-order solver for guided sampling.

3. We find that the multistep version is slightly better than the singlestep version.

Therefore, I suggest using the 2nd-order multistep DPM-Solver++, which is the above default settings.

**However, the 3rd-order method is still useful for unconditional sampling**, and it can achieve a better sample quality than the 2nd-order method for unconditional sampling. So I also provide a 3rd-order method in this PR.

References:
[1] DPM-Solver: https://arxiv.org/abs/2206.00927
[2] DPM-Solver++: https://arxiv.org/abs/2211.01095


Q3: We are still thinking about an API for schedulers that require two model evaluations. Will have a draft PR next week. Then we can add the other version of DPM once the API is finalised.

Great to know it! I can help to add the corresponding singlestep DPM-Solver at that time.

Q4: To differentiate it from the other scheduler could we call this DPMMultiStepScheduler ?

Good idea! I will change the name in the latter commits.

Q5: other modifications.

Thanks for your careful review! I will fix them in the latter commits.

LuChengTHU commented on src/diffusers/schedulers/scheduling_dpmsolver_discrete.py:
```diff
@@ -0,0 +1,471 @@
+                - (sigma_t * (torch.exp(h) - 1.0)) * D0
+                - (sigma_t * ((torch.exp(h) - 1.0) / h - 1.0)) * D1
+                - (sigma_t * ((torch.exp(h) - 1.0 - h) / h**2 - 0.5)) * D2
+            )
```
resolved in commit 588ce52, add comments

patil-suraj commented:
Thanks a lot for the detailed answer and addressing the review comments @LuChengTHU !  Good to merge for me :) 

@patrickvonplaten or @anton-l would be nice if you could also take quick look :)

patil-suraj commented on src/diffusers/utils/dummy_torch_and_accelerate_objects.py:
```diff
@@ -0,0 +1,467 @@
+# This file is autogenerated by the command `make fix-copies`, do not edit.
+# flake8: noqa
+
+from ..utils import DummyObject, requires_backends
+
+
+class ModelMixin(metaclass=DummyObject):
+    _backends = ["torch", "accelerate"]
+
+    def __init__(self, *args, **kwargs):
+        requires_backends(self, ["torch", "accelerate"])
+
+    @classmethod
+    def from_config(cls, *args, **kwargs):
+        requires_backends(cls, ["torch", "accelerate"])
+
+    @classmethod
+    def from_pretrained(cls, *args, **kwargs):
+        requires_backends(cls, ["torch", "accelerate"])
+
+
+class AutoencoderKL(metaclass=DummyObject):
+    _backends = ["torch", "accelerate"]
+
+    def __init__(self, *args, **kwargs):
+        requires_backends(self, ["torch", "accelerate"])
+
```
One last bit 😅

This file is not needed anymore. Could you 
- delete the file
- rebase to main 
- run `make fix-copies` 
- and push again 

then it should be good to merge  :)

LuChengTHU replied:
done :)

patrickvonplaten commented on src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:
```diff
@@ -0,0 +1,479 @@
+        solver_order (`int`, default `2`):
+            the order of DPM-Solver; can be `1` or `2` or `3`. We recommend to use `solver_order=2` for guided
+            sampling, and `solver_order=3` for unconditional sampling.
+        predict_x0 (`bool`, default `True`):
```
Could we maybe rename this variable?
So that it better matches: https://github.com/huggingface/diffusers/blob/08a6dc8a5840e0cc09e65e71e9647321ab9bb254/src/diffusers/schedulers/scheduling_ddpm.py#L223

BTW, it's definitely the better/correct choice to put this as a config argument as done here instead of having it be a function arg => We have an open PR to correct this for DDPM as well: https://github.com/huggingface/diffusers/pull/1155

```suggestion
        predict_epsilon (`bool`, default `False`):
```

LuChengTHU replied:
fixed in f54cf99

patrickvonplaten commented on src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:
```diff
@@ -0,0 +1,479 @@
+        beta_schedule: str = "linear",
+        trained_betas: Optional[np.ndarray] = None,
+        solver_order: int = 2,
+        predict_x0: bool = True,
```
```suggestion
        predict_epsilon: bool = False,
```

LuChengTHU replied:
fixed in f54cf99

patrickvonplaten commented on tests/test_scheduler.py:
```diff
@@ -549,6 +550,173 @@ def test_full_loop_with_no_set_alpha_to_one(self):
+            "beta_end": 0.02,
+            "beta_schedule": "linear",
+            "solver_order": 2,
+            "predict_x0": True,
```
```suggestion
            "predict_epsilon": False,
```

LuChengTHU replied:
fixed in f54cf99

patrickvonplaten commented on tests/test_scheduler.py:
```diff
@@ -549,6 +550,173 @@ def test_full_loop_with_no_set_alpha_to_one(self):
+                    self.check_over_configs(
+                        thresholding=True,
+                        sample_max_value=threshold,
+                        predict_x0=True,
```
```suggestion
                        predict_epsilon=False,
```

LuChengTHU replied:
fixed in f54cf99

patrickvonplaten commented on tests/test_scheduler.py:
```diff
@@ -549,6 +550,173 @@ def test_full_loop_with_no_set_alpha_to_one(self):
+    def test_solver_order_and_type(self):
+        for solver_type in ["dpm_solver", "taylor"]:
+            for order in [1, 2, 3]:
+                for predict_x0 in [True, False]:
```
```suggestion
                for predict_epsilon in [True, False]:
```

LuChengTHU replied:
fixed in f54cf99

patrickvonplaten commented on tests/test_scheduler.py:
```diff
@@ -549,6 +550,173 @@ def test_full_loop_with_no_set_alpha_to_one(self):
+        for solver_type in ["dpm_solver", "taylor"]:
+            for order in [1, 2, 3]:
+                for predict_x0 in [True, False]:
+                    self.check_over_configs(solver_order=order, solver_type=solver_type, predict_x0=predict_x0)
```
```suggestion
                    self.check_over_configs(solver_order=order, solver_type=solver_type, predict_epsilon=predict_epsilon)
```

LuChengTHU replied:
fixed in f54cf99

patrickvonplaten commented on tests/test_scheduler.py:
```diff
@@ -549,6 +550,173 @@ def test_full_loop_with_no_set_alpha_to_one(self):
+            for order in [1, 2, 3]:
+                for predict_x0 in [True, False]:
+                    self.check_over_configs(solver_order=order, solver_type=solver_type, predict_x0=predict_x0)
+                    sample = self.full_loop(solver_order=order, solver_type=solver_type, predict_x0=predict_x0)
```
```suggestion
                    sample = self.full_loop(solver_order=order, solver_type=solver_type, predict_epsilon=predict_epsilon)
```

LuChengTHU replied:
fixed in f54cf99

patrickvonplaten commented on src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:
```diff
@@ -0,0 +1,479 @@
+        D1_0, D1_1 = (1.0 / r0) * (m0 - m1), (1.0 / r1) * (m1 - m2)
+        D1 = D1_0 + (r0 / (r0 + r1)) * (D1_0 - D1_1)
+        D2 = (1.0 / (r0 + r1)) * (D1_0 - D1_1)
+        if self.config.predict_x0:
```
```suggestion
        if not self.config.predict_epsilon:
```

LuChengTHU replied:
fixed in f54cf99

patrickvonplaten commented on src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:
```diff
@@ -0,0 +1,479 @@
+        h, h_0 = lambda_t - lambda_s0, lambda_s0 - lambda_s1
+        r0 = h_0 / h
+        D0, D1 = m0, (1.0 / r0) * (m0 - m1)
+        if self.config.predict_x0:
```
```suggestion
        if not self.config.predict_epsilon:
```

LuChengTHU replied:
fixed in f54cf99

patrickvonplaten commented on src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:
```diff
@@ -0,0 +1,479 @@
+        alpha_t, alpha_s = self.alpha_t[prev_timestep], self.alpha_t[timestep]
+        sigma_t, sigma_s = self.sigma_t[prev_timestep], self.sigma_t[timestep]
+        h = lambda_t - lambda_s
+        if self.config.predict_x0:
```
```suggestion
        if not self.config.predict_epsilon:
```

LuChengTHU replied:
fixed in f54cf99

patrickvonplaten commented on src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:
```diff
@@ -0,0 +1,479 @@
+        Returns:
+            `torch.FloatTensor`: the converted model output.
+        """
+        if self.config.predict_x0:
```
```suggestion
        if not self.config.predict_epsilon:
```

LuChengTHU replied:
fixed in f54cf99

patrickvonplaten commented on src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:
```diff
@@ -0,0 +1,479 @@
+            sampling, and `solver_order=3` for unconditional sampling.
+        predict_x0 (`bool`, default `True`):
+            DPM-Solver is designed for both the noise prediction model (DPM-Solver, https://arxiv.org/abs/2206.00927)
+            with `predict_x0=False` and the data prediction model (DPM-Solver++, https://arxiv.org/abs/2211.01095) with
```
```suggestion
            with `predict_epsilon=True` and the data prediction model (DPM-Solver++, https://arxiv.org/abs/2211.01095) with
```

LuChengTHU replied:
fixed in f54cf99

patrickvonplaten commented on src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:
```diff
@@ -0,0 +1,479 @@
+        predict_x0 (`bool`, default `True`):
+            DPM-Solver is designed for both the noise prediction model (DPM-Solver, https://arxiv.org/abs/2206.00927)
+            with `predict_x0=False` and the data prediction model (DPM-Solver++, https://arxiv.org/abs/2211.01095) with
+            `predict_x0=True`. We recommend to use `predict_x0=True` and `solver_order=2` for guided sampling (e.g.
```
```suggestion
            `predict_epsilon=False`. We recommend to use `predict_epsilon=False` and `solver_order=2` for guided sampling (e.g.
```

LuChengTHU replied:
fixed in f54cf99

patrickvonplaten commented on src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:
```diff
@@ -0,0 +1,479 @@
+            stable-diffusion).
+        thresholding (`bool`, default `False`):
+            whether to use the "dynamic thresholding" method (introduced by Imagen, https://arxiv.org/abs/2205.11487).
+            For pixel-space diffusion models, you can set both `predict_x0=True` and `thresholding=True` to use the
```
```suggestion
            For pixel-space diffusion models, you can set both `predict_epsilon=False` and `thresholding=True` to use the
```

LuChengTHU replied:
fixed in f54cf99

patrickvonplaten commented on src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:
```diff
@@ -0,0 +1,479 @@
+            the ratio for the dynamic thresholding method. Default is `0.995`, the same as Imagen
+            (https://arxiv.org/abs/2205.11487).
+        sample_max_value (`float`, default `1.0`):
+            the threshold value for dynamic thresholding. Valid only when `thresholding=True` and `predict_x0=True`.
```
```suggestion
            the threshold value for dynamic thresholding. Valid only when `thresholding=True` and `predict_epsilon=False`.
```

LuChengTHU replied:
fixed in f54cf99

patrickvonplaten commented on src/diffusers/schedulers/scheduling_dpmsolver_multistep_flax.py:
```diff
@@ -0,0 +1,563 @@
+# Copyright 2022 TSAIL Team and The HuggingFace Team. All rights reserved.
```
The same proposed changes as in the PyTorch file

LuChengTHU replied:
fixed in https://github.com/huggingface/diffusers/commit/f54cf995f4106e835513073c75e38b26da3eea01

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:
```diff
@@ -59,7 +60,12 @@ def __init__(
         unet: UNet2DConditionModel,
         scheduler: Union[
-            DDIMScheduler, PNDMScheduler, LMSDiscreteScheduler, EulerDiscreteScheduler, EulerAncestralDiscreteScheduler
+            DDIMScheduler,
```
Super nice that it works out of the box with the existing API

patrickvonplaten commented on src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:
```diff
@@ -0,0 +1,479 @@
+
+    """
+
+    _compatible_classes = [
```
Super nice that you added those! This allows the scheduler to be loaded right away from the original repos.
For completeness could you also add vice-versa the `DPMSolverMultistepScheduler` to the compatible list in the following files:

https://github.com/huggingface/diffusers/blob/08a6dc8a5840e0cc09e65e71e9647321ab9bb254/src/diffusers/schedulers/scheduling_ddpm.py#L105
https://github.com/huggingface/diffusers/blob/08a6dc8a5840e0cc09e65e71e9647321ab9bb254/src/diffusers/schedulers/scheduling_ddim.py#L112
https://github.com/huggingface/diffusers/blob/08a6dc8a5840e0cc09e65e71e9647321ab9bb254/src/diffusers/schedulers/scheduling_euler_discrete.py#L71
https://github.com/huggingface/diffusers/blob/08a6dc8a5840e0cc09e65e71e9647321ab9bb254/src/diffusers/schedulers/scheduling_euler_ancestral_discrete.py#L70
https://github.com/huggingface/diffusers/blob/08a6dc8a5840e0cc09e65e71e9647321ab9bb254/src/diffusers/schedulers/scheduling_pndm.py#L91
https://github.com/huggingface/diffusers/blob/08a6dc8a5840e0cc09e65e71e9647321ab9bb254/src/diffusers/schedulers/scheduling_lms_discrete.py#L70

LuChengTHU replied:
fixed in dee238f

patrickvonplaten reviewed (approved):
Amazing addition @LuChengTHU! 

The API choices are great and it's amazing that it fits so well with the existing API of Stable Diffusion :-) 

Only left a couple of nits:
- suggestion to rename `predict_x0` to `predict_epsilon` for consistency 
- Add the DPM-Solver scheduler as a compatible scheduler to all other schedulers

LuChengTHU commented:
Hi @patrickvonplaten, thank you for the quick reviews! I've fixed all the modifications. The main difference is that I've changed the APIs for `predict_epsilon`, `algorithm_type`, and `solver_type` to a clearer way (and also changed the docs).

patil-suraj commented:
Thank you for addressing the comments, let's go!
